parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,421 @@
+        self.metrics['examples'] += len(batch['observations'])
+        self.metrics['loss'] += loss
+
+        self.backward(loss)
+        self.update_params()
+        #self.optimizer.step()
+        self.zero_grad()
+        #self.optimizer.zero_grad()
+
+        # # predictions
+        # with torch.no_grad():
+        #     b_input_ids, b_segment_ids, b_input_mask, _, _ = tensors
+        #     print(b_input_ids)
+        #     input(...)
+        #     b_start_logits, b_end_logits = self.model(b_input_ids,
+        #                                               b_segment_ids,
+        #                                               b_input_mask)
+
+        # predictions = []
+        # for start_logits, end_logits, input_ids in zip(b_start_logits,
+        #                                                b_end_logits,
+        #                                                b_input_ids):
+        #     prediction = self._get_prediction(start_logits, end_logits,
+        #                                       input_ids)
+        #     predictions.append(prediction)
+
+        # return Output(predictions)
+
+    def eval_step(self, batch):
+
+        segment_ids = torch.zeros(batch['text_vec'].size(),
+                                  dtype=torch.long).to(self.device)
+        input_mask = batch['text_vec'].clone()
+        input_mask[input_mask != self.dict.pad_idx] = 1
+
+        with torch.no_grad():
+            b_start_logits, b_end_logits = self.model(batch['text_vec'],
+                                                      segment_ids, input_mask)
+
+        predictions = []
+        for start_logits, end_logits, input_ids in zip(b_start_logits,
+                                                       b_end_logits,
+                                                       batch['text_vec']):
+            prediction = self._get_prediction(start_logits, end_logits,
+                                              input_ids)
+            predictions.append(prediction)
+
+        return Output(predictions)
+
+    def share(self):
+        """Share model parameters."""
+        shared = super().share()
+        shared['model'] = self.model
+        shared['metrics'] = self.metrics
+        return shared
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        TorchAgent.add_cmdline_args(parser)
+        parser = parser.add_argument_group('BERT Classifier Arguments')
+        parser.add_argument('--type-optimization',
+                            type=str,
+                            default='all_encoder_layers',
+                            choices=[
+                                'additional_layers', 'top_layer',
+                                'top4_layers', 'all_encoder_layers', 'all'
+                            ],
+                            help='which part of the encoders do we optimize '
+                            '(defaults to all layers)')
+        parser.add_argument(
+            "--max_text_length",
```

Comment:
how is this different than --truncate/--text-truncate/--label-truncate?

Our convventions use `--hyphens-instead-of-underscores`.